lib/imagery/iscatt_core.c

```diff
@@ -628,29 +628,32 @@ static void get_needed_bands(struct scCats *cats, int *b_needed_bands)
  */
 static void free_compute_scatts_data(int *fd_bands, struct rast_row *bands_rows,
                                      int n_a_bands, int *bands_ids,
-                                     int *fd_cats_rasts,
+                                     int *b_needed_bands, int *fd_cats_rasts,
                                      FILE **f_cats_rasts_conds, int n_a_cats)
 {
-    int i, band_id;
-
-    for (i = 0; i < n_a_bands; i++) {
-        band_id = bands_ids[i];
+    for (int i = 0; i < n_a_bands; i++) {
+        int band_id = bands_ids[i];
         if (band_id >= 0) {
             Rast_close(fd_bands[i]);
             G_free(bands_rows[band_id].row);
             G_free(bands_rows[band_id].null_row);
         }
     }
 
-    if (f_cats_rasts_conds)
-        for (i = 0; i < n_a_cats; i++)
-            if (f_cats_rasts_conds[i])
-                fclose(f_cats_rasts_conds[i]);
+    for (int i = 0; i < n_a_cats; i++)
+        if (f_cats_rasts_conds[i])
+            fclose(f_cats_rasts_conds[i]);
+
+    for (int i = 0; i < n_a_cats; i++)
+        if (fd_cats_rasts[i] >= 0)
+            Rast_close(fd_cats_rasts[i]);
 
-    if (fd_cats_rasts)
-        for (i = 0; i < n_a_cats; i++)
-            if (fd_cats_rasts[i] >= 0)
-                Rast_close(fd_cats_rasts[i]);
+    G_free(fd_bands);
+    G_free(bands_rows);
+    G_free(bands_ids);
+    G_free(b_needed_bands);
+    G_free(fd_cats_rasts);
+    G_free(f_cats_rasts_conds);
```

Comment:
Given that this removes the check and adds free, can you please confirm that you checked that this function is indeed used in this way? Updating/adding the doc may be helpful too since now you actually made it behave in a specific way rather than just "whatever the helper function happens to do that's it" which was the previous state.

Reply:
> Given that this removes the check and adds free, can you please confirm that you checked that this function is indeed used in this way? Updating/adding the doc may be helpful too since now you actually made it behave in a specific way rather than just "whatever the helper function happens to do that's it" which was the previous state.

This private function (`free_compute_scatts_data()`) is only used by `I_compute_scatts()` for cleaning up allocated data. If one looks at the code in total, not only the diff, it is at least for me pretty clear what is going on. The only change is that now all arrays are (explicitly) malloc'd and need to be freed explicitly by every call.